File: platforms/android/runtime/core/src/main/java/io/nativeblocks/runtime/api/provider/type/NativeTypeProvider.kt
package io.nativeblocks.runtime.api.provider.type

import java.util.concurrent.ConcurrentHashMap
import kotlin.reflect.KClass


/**
 * A provider for handling type converters for various data types.
 * This class maps specific types to their respective converters,
 * allowing easy conversion between objects and their string representations.
 */
internal class NativeTypeProvider {

    private val types = mutableMapOf<KClass<*>, INativeType<*>>()

    /**
     * Registers a type converter for a given type.
     *
     * @param T The type for which the converter is being provided.
     * @param type The `KClass` of the type.
     * @param converter The implementation of `INativeType` to handle conversion for the type.
     */
    fun <T : Any> provideTypeConverter(type: KClass<T>, converter: INativeType<T>) {
        types[type] = MemoizedNativeType(converter)
    }

    /**
     * Retrieves the type converter for a given type.
     *
     * @param T The type for which the converter is being retrieved.
     * @param type The `KClass` of the type.
     * @return The `INativeType` implementation for the requested type.
     * @throws NullPointerException If no converter has been registered for the requested type.
     */
    fun <T : Any> getTypeConverter(type: KClass<T>): INativeType<T> {
        try {
            @Suppress("UNCHECKED_CAST")
            return types[type] as INativeType<T>
        } catch (e: Exception) {
            throw NullPointerException("The ${type.qualifiedName} converter not provided.")
        }
    }
}

private class MemoizedNativeType<T : Any>(private val delegate: INativeType<T>) : INativeType<T> {

    private val cache = ConcurrentHashMap<String, T>()

    override fun toString(input: T?): String = delegate.toString(input)

    override fun fromString(input: String?): T {
        input ?: return delegate.fromString(null)
        return cache.getOrPut(input) { delegate.fromString(input) }
    }
}

File: platforms/ios/runtime/Sources/NativeblocksRuntime/api/provider/type/NativeTypeProvider.swift
import Foundation

/// A class responsible for managing and providing type converters for Native types.
/// This class allows registering and retrieving type converters for specific types.
internal class NativeTypeProvider {
    private var types: [String: Any] = [:]

    ///
    ///Registers a type converter for a specific type.
    ///
    ///- Parameters:
    ///  - type: The type for which the converter is being registered.
    ///  - converter: The converter that implements `INativeType` for the given type.
    ///
    func provideTypeConverter<T>(_ type: T.Type, converter: INativeType<T>) {
        let typeName = String(describing: type)
        types[typeName] = MemoizedNativeType(delegate: converter)
    }

    ///
    ///Retrieves the type converter for a specific type.
    ///
    ///- Parameter type: The type for which the converter is being retrieved.
    ///- Returns: The converter that implements `INativeType` for the given type.
    ///- Precondition: A converter for the specified type must have been registered using `provideTypeConverter`.
    ///- Postcondition: If no converter is found, a fatal error will occur.
    ///
    func getTypeConverter<T>(_ type: T.Type) -> INativeType<T> {
        let typeName = String(describing: type)
        if let serializer = types[typeName] as? INativeType<T> {
            return serializer
        } else {
            fatalError("The \(typeName) converter not provided.")
        }
    }
}

/// A converter is called once per property, per block, per frame update, so the
/// same string is decoded over and over. This caches `fromString` results.
private final class MemoizedNativeType<T>: INativeType<T> {

    private let delegate: INativeType<T>
    private var cache: [String: T] = [:]
    private let lock = NSLock()

    init(delegate: INativeType<T>) {
        self.delegate = delegate
        super.init()
    }

    override func toString(_ input: T?) -> String {
        return delegate.toString(input)
    }

    override func fromString(_ input: String?) -> T {
        guard let input else { return delegate.fromString(nil) }

        lock.lock()
        defer { lock.unlock() }
        if let cached = cache[input] { return cached }
        let value = delegate.fromString(input)
        cache[input] = value
        return value
    }
}
